engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionIndexIterator.java

```diff
@@ -21,53 +20,46 @@ final class RegionIndexIterator implements PrimitiveIterator.OfInt, Closeable {
      * @return the region index iterator
      */
     public static RegionIndexIterator of(final RowSet rowSet) {
-        return new RegionIndexIterator(rowSet.searchIterator());
+        // Note: we are doing an up-front check, with our implementation doing the follow-up hasNext check as part of
+        // the nextInt calculation. This allows us to be efficient and call SearchIterator#advance and
+        // SearchIterator#currentValue exactly once without needing to otherwise keep more complicated state handling
+        // if hasNext was on-demand.
+        final RowSet.SearchIterator sit = rowSet.searchIterator();
+        return new RegionIndexIterator(sit, nextRegion(sit, 0));
     }
 
+    private static final int DONE_REGION = -1;
+
     private final RowSet.SearchIterator sit;
-    private long key;
+    private int nextRegion;
+
+    private static int nextRegion(final RowSet.SearchIterator sit, final long key) {
+        return sit.advance(key)
+                ? RegionedColumnSource.getRegionIndex(sit.currentValue())
+                : DONE_REGION;
+    }
 
-    private RegionIndexIterator(final RowSet.SearchIterator sit) {
+    private RegionIndexIterator(final RowSet.SearchIterator sit, final int nextRegion) {
         this.sit = Objects.requireNonNull(sit);
-        this.key = 0;
+        this.nextRegion = nextRegion;
     }
 
     @Override
     public boolean hasNext() {
-        return key >= 0 && sit.advance(key);
+        return nextRegion >= 0;
```

Comment:
This is fine, but I think it might be clearer as 
```suggestion
        return nextRegion != DONE_REGION;
```